### PR TITLE
fix multi copy step data index

### PR DESCRIPTION
**Pull request by DreamWuGit**
this PR  can be part of #476 when doing multiple  copy step in buss mapping ,  for `TxLogFieldTag::Data` index should start from zero and consecutively increasing in multi copy log steps.  i.e.  total copies size is 66 bytes, data index should be monotone increasing by 1 from 0 across all three copy steps.  

| step      | data_start_index |copy range |
| ----------- | ----------- |------- |
| CopyStep1    | 0         | 0 --> 31    |
| CopyStep2   | 32       |32 --> 63  |
| CopyStep2   | 64       |64 -> 65    |.   

while old is as below though `Data` value is correct !  
| step     | data_start_index |copy range |
| ----------- | ----------- |------- |
| CopyStep1    | 0       | 0 --> 31    |
| CopyStep2   | 0       | 0 --> 31   |
| CopyStep2   | 0       | 0 --> 31    |.     

found in issue #511

### CODE REVIEW

File: zkevm-circuits/src/evm_circuit/witness.rs [hunk 1]
```diff
@@ -509,9 +509,9 @@ pub enum Rw {
         log_id: u64, // pack this can index together into address?
         field_tag: TxLogFieldTag,
         // topic index (0..4) if field_tag is TxLogFieldTag:Topic
-        // byte index (0..32) if field_tag is TxLogFieldTag:Data
+        // byte index if field_tag is TxLogFieldTag:Data
         // 0 for other field tags
-        index: u8,
+        index: usize,
```
Comment (z2trillion) [inline zkevm-circuits/src/evm_circuit/witness.rs:514]: On ~line 759, can you change 
```
Self::TxLog { log_id, index, .. } => {
     Some((U256::from(*index as u64) + (U256::from(*log_id) << 8)).to_address())
}
``` 
to
```
Self::TxLog { log_id, index, .. } => {
     Some((U256([*index as u64, *log_id, 0, 0]).to_address())
}
```

**Review by icemelon — COMMENTED** [review]
should add a test case that could cover the new fix code

File: zkevm-circuits/src/evm_circuit/witness.rs (hunk 1 above)
Comment (DreamWuGit) [reply zkevm-circuits/src/evm_circuit/witness.rs:514]: updated  in https://github.com/appliedzkp/zkevm-circuits/pull/510/commits/245e0688f9166fe88c01d03edcc1b6b48fc5fd17 ,however, I think I will refactor this line soon in coming PR as I think field tag should also place here if we consider sorting.

**Comment by DreamWuGit** [comment]
> 



> should add a test case that could cover the new fix code

tests of `copy_to_log_multi_step` cover this,  the incoming log test with multiple steps can indirect check this !

File: zkevm-circuits/src/evm_circuit/witness.rs (hunk 1 above)
Comment (DreamWuGit) [reply zkevm-circuits/src/evm_circuit/witness.rs:514]: realize that change to  `Some(U256([*index as u64, *log_id, 0, 0]).to_address())` will cause lookup error, I prefer to remain it for now as it is not closely related to this PR's target .

.

File: bus-mapping/src/circuit_input_builder/execution.rs [hunk 2]
```diff
@@ -144,7 +144,7 @@ pub enum CopyDetails {
     Code(U256),
     /// The bytes are being copied to a Log.
     /// Call's state change's persistance and tx_id are provided.
-    Log((bool, usize)),
+    Log((bool, usize, usize)),
```
[the remark this answers is not in the document]
Comment (ed255) [reply bus-mapping/src/circuit_input_builder/execution.rs:147]: It seems this third parameter is `data_start_index`.  Why not use `StepAuxiliaryData::dst_addr` for that?

File: zkevm-circuits/src/evm_circuit/execution/copy_to_log.rs [hunk 3]
```diff
@@ -68,7 +71,7 @@ impl<F: Field> ExecutionGadget<F> for CopyToLogGadget<F> {
                 cb.tx_log_lookup(
                     tx_id.expr(),
                     TxLogFieldTag::Data,
-                    i.expr(),
+                    data_start_index.expr() + i.expr(),
```
Comment (ed255) [inline zkevm-circuits/src/evm_circuit/execution/copy_to_log.rs:74]: I see that in line 90 you do the following constraints for the next step when the copy is not yet finished:
- next src_addr
- next bytes_left
- next src_addr_end

Shouldn't it also constraint that this parameter is the same?:
- next tx_id

And also the next `data_start_index` is?:
- next data_start_index = data_start_index +  MAX_COPY_BYTES

Please, let me know if I missed something.

Comment (DreamWuGit) [reply zkevm-circuits/src/evm_circuit/execution/copy_to_log.rs:74]: Cool,  added in https://github.com/appliedzkp/zkevm-circuits/pull/510/commits/e592c5e920eae2677e69dae001b66b1f14cee716 and also constrain data start index start with zero in first copy step, thanks for catching !

**Review by ed255 — APPROVED** [review]
LGTM! Thanks for the fix :)

**Review by icemelon — APPROVED** [review]
LGTM

**Pull request merged**